### PR TITLE
C# logging API

### LINKED ISSUE #1991: Provide logging hook for C#
Currently all the logs are printed out to Console.WriteLine(). Make logging destination configurable.

Alternatively, just use one of the commonly used .NET logging frameworks (like log4net).

### PR DESCRIPTION
**Pull request by jtattermusch**
This is based on #2645, which needs to be merged first.

Provides logging API that can redirect logs to a custom user-specified logger. Also gets rid of ugly Console.WriteLine in Grpc.Core.

Fixes #1991

### CODE REVIEW

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs [hunk 1]
```diff
@@ -212,7 +215,7 @@ public void StartDuplexStreamingCall(Metadata headers)
         }
 
         /// <summary>
-        /// Sends a streaming request. Only one pending send action is allowed at any given time.
+        /// Sends a streaming request. Only one pending send action is allologger any given time.
```
Comment (stanley-cheung) [inline src/csharp/Grpc.Core/Internal/AsyncCall.cs]: typo?

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs [hunk 2]
```diff
@@ -221,7 +224,7 @@ public void StartSendMessage(TRequest msg, AsyncCompletionDelegate<object> compl
         }
 
         /// <summary>
-        /// Receives a streaming response. Only one pending read action is allowed at any given time.
+        /// Receives a streaming response. Only one pending read action is allowed at any givelogger.
```
Comment (stanley-cheung) [inline src/csharp/Grpc.Core/Internal/AsyncCall.cs]: typo?

File: src/csharp/Grpc.Core/Internal/Timespec.cs [hunk 3]
```diff
@@ -74,54 +87,164 @@ public static Timespec InfFuture
...
+        /// For Timespec.InfPast or if timespec is before the lowest representable DateTime, DateTime.MinValue is returned.
+        /// 
+        /// Unless DateTime.MaxValue or DateTime.MinValue is returned, the resulting DateTime is always in UTC
+        /// (DateTimeKind.Utc)
+        /// </summary>
+        public DateTime ToDateTime()
         {
-            if (timeout == Timeout.InfiniteTimeSpan)
+            Preconditions.CheckState(tv_nsec >= 0 && tv_nsec < NanosPerSecond);
+            Preconditions.CheckState(clock_type == GPRClockType.Realtime);
```
Comment (stanley-cheung) [inline src/csharp/Grpc.Core/Internal/Timespec.cs]: I am not sure how they are related, but there had been recent work from C core related to these. See #2629. Nothing related to C-sharp had been done in that PR. Was just wondering if that change applies here?

**Comment by stanley-cheung** [comment]
The rest LGTM

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs (hunk 1 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/AsyncCall.cs]: Yes, something really weird was going on with refactor feature in Monodevelop IDE. Thanks for spotting this.

File: src/csharp/Grpc.Core/Internal/Timespec.cs (hunk 3 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/Timespec.cs]: Yes, that is already taken into account (I do an explicit conversion from monotonic to realtime). thanks for pointing that out anyway.

**Pull request merged**